Port to imgui-java bindings

nopjmp opened the pull request:
Looking into switching to imgui-java to get the stability of using the main imgui library until Kotlin Graphics becomes more stable.

nopjmp committed:
WIP: add in imgui-java to the gradle configuration

Not fully tested, but it should be included properly for every OS.

zeroeightysix committed:
Removed kotlin-graphics/imgui

zeroeightysix committed:
Move non-hud stuff from KamiHud to new KamiImgui

zeroeightysix committed:
Added ImguiDSL to replace old k-g/imgui's DSL

zeroeightysix committed:
wip: Ported a bunch of classes to imgui bindings

zeroeightysix committed:
wip: Ported more stuff, expanded DSL

zeroeightysix committed:
wip: PinnableWidget and Settings, partially CompiledText

nopjmp committed:
wip: Finished up CompiledText and PrepHandler

Switched up how Colour works internally to use the dedicated scalar arrays

Added in helper for floatRGBA to Colour conversion.

zeroeightysix committed:
fix: `wrapImInt` to update underlying property if dirty

nopjmp committed:
wip: fixed up TargetSupplier

nopjmp committed:
wip: GraphPinnableWidget, EnabledWidgets, and more wrapIm* helper modifications

nopjmp committed:
wip: Continue with partial Widget porting

Information, PlayerPinnableWidget, TextPinnableWidget, VoidContextMenu

fix up modules

nopjmp committed:
wip: more partial re-implementations for Settings and Modules

Hacks need to be removed as they are using the internal API.

nopjmp committed:
wip: partial implementation of Wizard

nopjmp committed:
wip: finish the last of the API porting in KamiConfig and MixinRebuildTask

nopjmp committed:
wip: fix build configuration to not throw out javassist.

Running it segfaults somewhere in libjvm.so. Something isn't right.

nopjmp commented:
```
Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  imgui.ImGui.nInitJni()V+0
j  imgui.ImGui.<clinit>()V+199
v  ~StubRoutines::call_stub
j  imgui.gl3.ImGuiImplGl3.setupBackendCapabilitiesFlags()V+0
j  imgui.gl3.ImGuiImplGl3.init(Ljava/lang/String;)V+5
j  imgui.gl3.ImGuiImplGl3.init()V+2
j  me.zeroeightsix.kami.gui.KamiImgui.<clinit>()V+40
```

Looks like it's crashing in the initialization code. 1.80 of imgui-java is out there, but the binding part of the code should be more of less the same. The only change was a function that got renamed, I think it was openPopupContextItem or something along those lines.

nopjmp committed:
wip: use ByteStreams for Java 8 compatibility

Attempt to rearrange the init to follow the Application java-imgui example.

nopjmp committed:
wip: update API usage and center Wizard properly.

nopjmp committed:
wip: add in a fonts list and fix graphical problems due to bad font rendering

nopjmp committed:
wip: oops forgot to commit this hunk

nopjmp committed:
wip: font madness, Wizard now completes, NBTScreen closes when X is pressed.

Wizard still needs some fixes for the theme selector

zeroeightysix committed:
fix: Use temporary `ImInt` when editing colour

zeroeightysix committed:
fix: Help marker shows only a question mark until hovered

zeroeightysix committed:
fix: Enum types's imgui interface returned the wrong type

zeroeightysix committed:
fix: Out of bounds exception when turning on rainbow mode

zeroeightysix committed:
fix: Initialise imgui state after setting up fonts

This must be done as the font atlas is built upon the `init` call.

zeroeightysix committed:
fix: Selecting themes

The combo didn't work properly as the underlying property is updated only after the `block` executes. We pass the `ImInt` to the block now, and use that instead.

zeroeightysix committed:
fix: Use pOpen-less method when no open bool is passed

This makes it so that there is no (useless) close button on the window anymore.

nopjmp committed:
fix: Player and Text PinnableWidgets can now render.

There is still a little funkiness with the title being null and crashes from the VoidContextMenu

nopjmp committed:
fix: PinnableWidgets mostly work now

Inventory draws on Player for some reason and I have no idea why yet.

nopjmp committed:
fix: widget name must be unique and not null or empty

nopjmp committed:
fix: minimize is breaking the shadowJar dependencies of imgui-java

We could revisit this if a ~14MB jar file is too large.

Most of this is native libraries.

zeroeightysix committed:
fix: Wrong variable used when editing variable parts

zeroeightysix committed:
wip: Implement `charButton`, almost

char is not centered yet and there's some unexplainable residual spacing between the char button and next element

zeroeightysix committed:
wip: Added methods to the DSL that I forgot in the previous commit

oops
build failing :(

zeroeightysix committed:
fix: Active modules list width

with `getWindowWidth` it would infinitely expand to the right

zeroeightysix committed:
wip: Store char queue for charButton to use

zeroeightysix committed:
fix: Editing a string part in HUD elements was fixed-length

zeroeightysix committed:
chore: reformat

zeroeightysix committed:
fix: Alignment in `charButton`

zeroeightysix committed:
fix: Minecraftia font rendering

zeroeightysix committed:
wip: Turn docking on by default

zeroeightysix committed:
fix: `wrapImString` makes resizable string

zeroeightysix committed:
fix: delayed `Digits` setting

another case where the wrapped object is used instead of the wrapper

SpaiR commented on src/main/kotlin/me/zeroeightsix/kami/gui/KamiImgui.kt:
```diff
@@ -95,6 +95,10 @@ object KamiImgui {
             else -> 110
         }
+
+        ImGui.getIO().addConfigFlags(1 shl 6) // 1<<6 should be the DockingEnable flag
```
It is actually public https://github.com/SpaiR/imgui-java/blob/master/example/src/main/java/Main.java#L27

zeroeightysix replied:
That's bizarre. IntelliJ didn't seem to find it.

zeroeightysix committed:
fix: Use public `DockingEnable` flag

zeroeightysix committed:
fix: weird spacing between parts in imgui rendering mode

zeroeightysix committed:
fix: '(empty)' message in HUD window was invisible

zeroeightysix committed:
fix: 'Remove' menu item on parts are tied to the part

zeroeightysix committed:
fix: useless combo when editing graph variable

zeroeightysix committed:
feat: Removed autoresize and child from Baritone settings window

this makes it slightly more usable, and doesn't limit resizing.

zeroeightysix committed:
fix: disallow opening kami GUI when kami GUI is already open

zeroeightysix committed:
fix: columns not disabled after drawing target editing interface

zeroeightysix committed:
fix: text fields in MWE using non-resizeable strings

zeroeightysix committed:
fix: `Module alignment` only affects module selectables

zeroeightysix committed:
fix: use `imgui` extension field for activemodules filter

This sets the resize factor and resizeable flag, which the standard constructor does not do

zeroeightysix committed:
fix: remove sodium from build.gradle.kts

crashes when enabling ESP

zeroeightysix committed:
fix: the entire wizard

zeroeightysix committed:
fix: border offset in wizard

oops might have missed this

zeroeightysix committed:
Downgrade Java back to 1.8
